flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/ScalarOperatorGens.scala

```diff
@@ -403,47 +445,71 @@ object ScalarOperatorGens {
            |    equals(${right.resultTerm}.getBinarySection());
            |}
            |""".stripMargin
-      GeneratedExpression(resultTerm, nullTerm, code, resultType)
+      wrapExpressionIfNonEq(
```

Comment:
Similarly, here we cannot call `wrapExpressionIfNonEq` directly.

I've composed a test case to reproduce the issue.
```scala
  @Test
  def debugging(): Unit = {
    //You can set the class access modifier to public temporarily for test 
    val rawType = new RawType[LogicalTypesTest.Human](
      classOf[LogicalTypesTest.Human],
      new KryoSerializer[LogicalTypesTest.Human](
        classOf[LogicalTypesTest.Human],
        new ExecutionConfig))
    val className = "org.apache.flink.table.types.LogicalTypesTest$Human"
    val serializerString = rawType.asSerializableString
    tEnv.executeSql(
      s"""
         |create table test_in1 (
         |  content raw('$className', '$serializerString')
         |) with ('connector' = 'filesystem', 'format' = 'raw', 'path' = '/tmp/test-dir-in-1')
         |""".stripMargin)
    tEnv.executeSql(
      s"""
         |create table test_in2 (
         |  content raw('$className', '$serializerString')
         |) with ('connector' = 'filesystem', 'format' = 'raw', 'path' = '/tmp/test-dir-in-2')
         |""".stripMargin)
    tEnv.executeSql(
      s"""
         |create table test_out (
         |  content raw('$className', '$serializerString')
         |) with ('connector' = 'filesystem', 'format' = 'raw', 'path' = '/tmp/test-dir-out')
         |""".stripMargin)
    tEnv
      .executeSql(
        """
          |insert into test_out
          |select c2 from (
          |  select in1.content as c1, in2.content as c2
          |  from test_in1 in1 join test_in2 in2
          |  on in1.content <> in2.content
          |) tmp
          |""".stripMargin)
      .await()
  }
```
The correct and wrong generated non-equi condition are as follows:
```java
// correct
boolean isNull$5 = isNull$0 || isNull$3;
          boolean result$5 = false;
          if (!isNull$5) {
            field$2.ensureMaterialized(typeSerializer$1.getInnerSerializer());
            field$4.ensureMaterialized(typeSerializer$1.getInnerSerializer());
            result$5 =
              !field$2.getBinarySection().
              equals(field$4.getBinarySection());
          }
          
          return result$5;
// wrong
boolean isNull$5 = isNull$0 || isNull$3;
          boolean result$5 = false;
          if (!isNull$5) {
            field$2.ensureMaterialized(typeSerializer$1.getInnerSerializer());
            field$4.ensureMaterialized(typeSerializer$1.getInnerSerializer());
            result$5 =
              field$2.getBinarySection().
              equals(field$4.getBinarySection());
          }
          
          return (!result$5);
```